frame/lottery/src/lib.rs

```diff
@@ -0,0 +1,298 @@
+
+	/// The manager origin.
+	type ManagerOrigin: EnsureOrigin<Self::Origin>;
+
+	/// The max number of calls available in a single lottery.
+	type MaxCalls: Get<usize>;
+
+	/// Weight information for extrinsics in this pallet.
+	type WeightInfo: WeightInfo;
+}
+
+type Index = u32;
+
+// Any runtime call can be encoded into two bytes which represent the pallet and call index.
+// We use this to uniquely match someone's incoming call with the calls configured for the lottery.
+type CallIndex = (u8, u8);
+
+decl_storage! {
+	trait Store for Module<T: Trait> as Lottery {
+		LotteryIndex: Index;
+		/// The configuration for the current lottery.
+		Lottery: Option<LotteryConfig<T::BlockNumber, BalanceOf<T>>>;
+		/// Users who have purchased a ticket. (Lottery Index, Tickets Purchased)
+		Participants: map hasher(twox_64_concat) T::AccountId => (Index, Vec<CallIndex>);
+		/// Total number of tickets sold.
+		TicketsCount: Index;
+		/// Each ticket's owner.
+		///
+		/// May have residual storage from previous lotteries. Use `TicketsCount` to see which ones
+		/// are actually valid ticket mappings.
+		Tickets: map hasher(twox_64_concat) Index => Option<T::AccountId>;
+	}
+}
+
+decl_event!(
+	pub enum Event<T> where
+		<T as frame_system::Trait>::AccountId,
+		Balance = BalanceOf<T>,
+		Call = <T as Trait>::Call,
+	{
+		/// A lottery has been started!
+		LotteryStarted,
+		/// A winner has been chosen!
+		Winner(AccountId, Balance),
+		/// A ticket has been bought!
+		TicketBought(AccountId, Call),
+	}
+);
+
+decl_error! {
+	pub enum Error for Module<T: Trait> {
+		/// An overflow has occurred.
+		Overflow,
+		/// A lottery has not been configured.
+		NotConfigured,
+		/// A lottery has not started.
+		NotStarted,
+		/// A lottery is already in progress.
+		InProgress,
+		/// A lottery has already ended.
+		AlreadyEnded,
+		/// The call is not valid for an open lottery.
+		InvalidCall,
+		/// You are already participating in the lottery with this call.
+		AlreadyParticipating,
+		/// Too many calls for a single lottery.
+		TooManyCalls,
+		/// Failed to encode calls
+		EncodingFailed,
+	}
+}
+
+#[derive(Encode, Decode, Default, RuntimeDebug)]
+pub struct LotteryConfig<BlockNumber, Balance> {
+	/// Index of this lottery
+	index: Index,
+	/// Price per entry.
+	price: Balance,
+	/// Starting block of the lottery.
+	start: BlockNumber,
+	/// Ending block of the lottery.
+	end: BlockNumber,
+	/// Payout block of the lottery.
+	payout: BlockNumber,
+	/// Calls available this lottery.
+	calls: Vec<CallIndex>,
+}
+
+decl_module! {
+	pub struct Module<T: Trait> for enum Call where origin: T::Origin, system = frame_system {
+		/// TODO: Expose Constants
+
+		fn deposit_event() = default;
+
+		#[weight = T::WeightInfo::setup_lottery(calls.len() as u32)]
+		fn setup_lottery(
+			origin,
+			price: BalanceOf<T>,
+			start: T::BlockNumber,
+			end: T::BlockNumber,
+			payout: T::BlockNumber,
+			calls: Vec<<T as Trait>::Call>,
```

Comment:
I'd actually suggest we hard-code this value as a trait -- instead of passing the calls.

This would allow easy parameterization. So we don't require people participating to only to be able to do `vote(50)`, but can vote with whatever values they have. It's also more stable -- in case if the encoding of calls changed.

Reply:
@sorpaas I don't exactly follow here.

One issue with programming it as a trait is that updating the lotteries will be much more difficult. Maybe that is not a big deal, but i would want to be convinced that this flexibility would not be right for this lottery system.

As for the parameters, they def do NOT need to match. You may initiate the lottery with `vote(50)`, but the only thing recorded and checked when trying to allow a lottery ticket purchase is that `vote(..)` is called.

So I think this is not an issue, but let me know if I am missing something.

Reply:
> As for the parameters, they def do NOT need to match. You may initiate the lottery with vote(50), but the only thing recorded and checked when trying to allow a lottery ticket purchase is that vote(..) is called.

Ah I see. But in this case it's not possible to specify a lower limit, right? Someone can create millions of account and do `vote(0.0000001)` to gain a million ticket. I read your document that the prevention to this is the transaction fees and account creation costs, but I wonder whether that alone would be enough. The thing is, the abuser will simply need to ensure that the transaction fees and account creation costs are lower than the reward it gets.

> One issue with programming it as a trait is that updating the lotteries will be much more difficult. Maybe that is not a big deal, but i would want to be convinced that this flexibility would not be right for this lottery system.

I was imagining that the lottery will be invoked often, with not-too-high amount each time, but with generally the same amount each time. So I was thinking the calls will not be changed nearly at all -- we'd always want lottery for governance, and it'll probably stay that way.

Reply:
@sorpaas I dont think that we can be so granular with the calls without hardcoding specific logic related to specific pallets.

For example, even if we support a check that vote must lock more than 50 tokens, nothing would then stop the user from changing their vote to a lower amount later.

As for making many accounts, in general, you can avoid this by not pre-funding the lottery. i.e. every lottery ticket will be negative expected value. But I suspect at small scales this will not really be an issue. And if there is an issue, it is something that can be addressed more specifically based on whats going on.